bsms.mediawiki

```diff
@@ -0,0 +1,176 @@
+* A QR code.
+* Other formats.
+
+The flexibility in the data format allows each Signer to customize the User Experience based on its respective capabilities.
+
+====Key Derivation====
+The key derivation function is [https://tools.ietf.org/html/rfc2898 PBKDF2], with PRF = SHA512. Specifically:
+
+<tt>DK = PBKDF2(PRF, Password, Salt, c, dkLen)</tt>
+
+Whereas:
+
+* PRF = SHA512
+* Password = "No SPOF"
+* Salt = <tt>TOKEN</tt>
+* c = 2048
+* dkLen = 256
+* DK = Derived <tt>ENCRYPTION_KEY</tt>
+
+====Encryption Scheme====
+The encryption scheme is [https://tools.ietf.org/html/rfc3686 AES-256-CTR].
+
+<tt>Ciphertext = AES-256-CTR-Encrypt(Plaintext, Key, IV)</tt>
+
+<tt>Plaintext = AES-256-CTR-Decrypt(Ciphertext, Key, IV)</tt>
+
+<tt>IV = HMAC-SHA256(HMAC_Key, TOKEN || Data)</tt>
+
+Whereas:
+* Key = <tt>ENCRYPTION_KEY</tt>
+* HMAC_Key = SHA256(<tt>ENCRYPTION_KEY</tt>)
+* Data = the plaintext data, i.e. the entire key record in round 1 and the entire descriptor record in round 2
+
+The IV is to be sent along with the key and descriptor record, as specified above. Because the IV is an HMAC over the entire plaintext, this is essentially an [https://en.wikipedia.org/wiki/Authenticated_encryption#Encrypt-and-MAC_(E&M) Encrypt-and-MAC] form of authenticated encryption.
+
+==QR Codes==
+For signers that use QR codes to transmit data, key and descriptor records can be converted to QR codes, following [https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md the BCR standard].
+
+Also refer to [https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-015-account.md UR Type Definition for BIP44 Accounts] and [https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-010-output-desc.md UR Type Definition for Bitcoin Output Descriptors] for more details.
+
+==Security==
+
+This proposal introduces two layers of protection. The first one is a temporary, secret token, used to encrypt the two rounds of communication between the Signer and the Coordinator. The second one is through the descriptor checksum and visual inspection of the descriptor itself.
+
+The token is only needed during the setup phase, and can be safely thrown away afterwards. The token does not guarantee that the Signer membership set is not modified, since that depends on the overall security of all parties in the setup, but it can make it significantly harder for an attacker to do so.
+
+There are three ways an attacker can modify the membership set: by changing an existing member, by removing an existing member, or by adding a new member.
+
+For the first two methods, one of the Signers will be able to detect that its membership has been changed or removed, and reject the final descriptor. Thus, it is vital that all participating Signers check that their membership is intact in the descriptor. Even one Signer failing to check for its membership means that the setup could be compromised.
```

Comment:
Suppose the signer detected that it is not included in the descriptor record they got from the Coordinator.

What should they do ? Should Coordinator or other participants delay generating the receive addresses (and using it as a payment destination) with the newly created descriptor record until all signers indicate they found their key in it ?

How signers should communicate this 'I has verified the descriptor record' ? By other channels, or maybe there could be an optional 3rd round where signers sign the final descriptor record ?

Reply:
> Should Coordinator or other participants delay generating the receive addresses (and using it as a payment destination) with the newly created descriptor record until all signers indicate they found their key in it?

They definitely should, good idea to add this in.  I imagine they could all communicate OOB as a check. But maybe it'd be better to build it in to the protocol.

Reply:
I think it's looking pretty great now. I don't know if a 3rd round is entirely necessary.  The signers already have a method of contacting each other if they are able to pass PSBTs to each other (and the coordinator passing TOKENs), so this wouldn't be entirely unreasonable.

Reply:
I'm also leaning towards not having an explicit round 3 for the reason @Rspigler stated.

Also, as part of round 2, we currently have this line:

> "Except for the position, all participating Signers should be able to display the same confirmation."

Which is sort of implying the "optional round 3" here. Perhaps we can make the language a bit stronger, e.g., "participating Signers must check with each other that everyone has verified the descriptor and has the same confirmation".